cmd/clusterctl/pkg/client/init.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"github.com/pkg/errors"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/cluster"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/config"
+)
+
+// Init a management cluster by adding the requested list of providers.
+
+func (c *clusterctlClient) Init(kubeconfig string, coreProvider string, bootstrapProviders, infrastructureProviders []string, targetNameSpace, watchingNamespace string, force bool) ([]Components, bool, error) {
+	// gets access to the management cluster
+	cluster, err := c.clusterClientFactory(kubeconfig)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// ensure the custom resource definitions required by clusterctl are in place
+	if err := cluster.ProviderInventory().EnsureCustomResourceDefinitions(); err != nil {
+		return nil, false, err
+	}
+
+	// checks if the cluster already contains a Core provider.
+	// if not we consider this the first time init is executed, and thus we enforce the installation of a core provider (if not already explicitly requested by the user)
+	firstRun := false
+	currentCoreProvider, err := cluster.ProviderInventory().GetDefaultProviderName(clusterctlv1.CoreProviderType)
+	if err != nil {
+		return nil, false, err
+	}
+	if currentCoreProvider == "" && coreProvider == "" {
+		firstRun = true
+		coreProvider = config.ClusterAPIName
+	}
```

Comment:
I'm wondering if we could make this into its own method and remove the need to return more than 2 parameters

Reply:
We can do in a follow-up PR when there will be a client for ProviderInventory() as well
I'm taking note in the tracking issue ATM